[google_maps_flutter] Support for Ground Overlay

jokerttu opened the pull request:
Adds Ground Overlay support for google_maps_flutter plugin.

Fixes https://github.com/flutter/flutter/issues/26479

[Android](https://developers.google.com/maps/documentation/android-sdk/groundoverlay):
<img src="https://github.com/user-attachments/assets/1b1b4b9b-0260-49c4-87ed-4de29f4741db" width=40%>

[iOS](https://developers.google.com/maps/documentation/ios-sdk/overlays):
<img src="https://github.com/user-attachments/assets/fd24e73d-70a0-4972-b7c2-595acc79c2e9" width=40%>

[Web](https://developers.google.com/maps/documentation/javascript/groundoverlays):
<img src="https://github.com/user-attachments/assets/06015fa8-6cc4-46ad-81f1-95966015d338" width=60%>

Documentation for groundOverlays parameter on app-facing plugin:
<img width="736" alt="documentation" src="https://github.com/user-attachments/assets/81bef3cb-2167-44fc-9c44-b45399ed4d30" />


## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I [linked to at least one issue that this PR fixes] in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style], or this PR is [exempt from CHANGELOG changes].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[Discord]: https://github.com/flutter/flutter/blob/master/docs/contributing/Chat.md
[linked to at least one issue that this PR fixes]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#overview
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#version
[following repository CHANGELOG style]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#changelog-style
[exempt from CHANGELOG changes]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#changelog
[test-exempt]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#tests

jokerttu commented on packages/google_maps_flutter/google_maps_flutter/example/integration_test/src/tiles_inspector.dart:
```diff
@@ -22,7 +21,7 @@ void main() {
 
 void runTests() {
-  const double floatTolerance = 1e-8;
+  const double floatTolerance = 1e-6;
```
Added bigger tolerance as on Android and iOS tests failed as diff was bigger than 1e-8

jokerttu commented on packages/google_maps_flutter/google_maps_flutter/example/integration_test/src/tiles_inspector.dart:
```diff
@@ -208,269 +207,574 @@ void runTests() {
-          Color.fromARGB(255, 0, 0, 191),
-          0.6,
+      );
+      await tester.pumpAndSettle(const Duration(seconds: 3));
```
Without this line tests timeouted on Web platform, and we were not able to test ground overlay below.

ditman replied:
Were they timing out in all tests, or wasm specifically? I know I had to revisit a bunch of plugins and add a few `pump`s and `pumpAndSettle` when enabling the wasm tests. See:

* https://github.com/flutter/packages/pull/8111

jokerttu commented on packages/google_maps_flutter/google_maps_flutter/example/integration_test/src/tiles_inspector.dart:
```diff
@@ -208,269 +207,574 @@ void runTests() {
-          Color.fromARGB(255, 255, 0, 0),
-          1,
+      );
+      await tester.pumpAndSettle(const Duration(seconds: 3));
```
Without this line heatmap tests timeouted on Web platform, and we were not able to test ground overlay below.

ditman replied:
(Same as above, resolving)

jokerttu commented on packages/google_maps_flutter/google_maps_flutter/example/integration_test/src/tiles_inspector.dart:
```diff
@@ -208,269 +207,574 @@ void runTests() {
-      expect(wll1.weight, wll2.weight);
-      expect(wll1.point.latitude, moreOrLessEquals(wll2.point.latitude));
-      expect(wll1.point.longitude, moreOrLessEquals(wll2.point.longitude));
+  group('Heatmaps', () {
```
Added Heatmap tests to own grounp, like tile overlay tests and ground overlay tests.

jokerttu commented on packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/platform_interface/google_maps_inspector_platform.dart:
```diff
@@ -117,11 +117,7 @@ abstract class GoogleMapsInspectorPlatform extends PlatformInterface {
-      'supportsGettingHeatmapInfo() has not been implemented.',
-    );
-  }
+  bool supportsGettingHeatmapInfo() => false;
```
This change is not ground overlay related, but as this method was not implemented for web platform, integration tests for web were failing because of this.
`supportsGettingHeatmapInfo() has not been implemented.`
We could have added implementation for this to web platform, but instead having this defaults to false, seems better solution.

ditman replied:
Sounds good to me, this is exactly what this "capability check" is supposed to do.

stuartmorgan replied:
Yes, this was a mistake; "supports"-style interface methods should default to false instead of throwing. Thanks for fixing it!

jokerttu commented on packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/platform_interface/google_maps_inspector_platform.dart:
```diff
@@ -132,6 +128,20 @@ abstract class GoogleMapsInspectorPlatform extends PlatformInterface {
   }
 
+  /// If the platform supports getting information about ground overlays.
+  bool supportsGettingGroundOverlayInfo() => false;
```
This follows same pattern as heatmap implementation.
Having this method could be unnecessary, as all platforms support getting groung operlay info at the moment.
We can remove this method seen valid change.

ditman replied:
I like that we're adding these capability checks to plugins, we should do this more.

stuartmorgan replied:
Yep, doing this more is actually our written policy now :)

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/example/ios14/ios/RunnerTests/GoogleMapsGroundOverlayControllerTests.m:
```diff
@@ -0,0 +1,155 @@
+@import GoogleMaps;
+
+#import <OCMock/OCMock.h>
+#import <google_maps_flutter_ios/FGMGroundOverlayController_Test.h>
```
Per style guide, this goes before all other imports, in its own section.

illuminati1911 replied:
Done.

stuartmorgan replied:
It should be before *all* other imports; see https://google.github.io/styleguide/objcguide.html#order-of-includes

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/example/ios14/ios/RunnerTests/GoogleMapsGroundOverlayControllerTests.m:
```diff
@@ -0,0 +1,155 @@
+@implementation GoogleMapsGroundOverlayControllerTests
+
+/// Returns GoogleMapGroundOverlayController object instantiated with position and a mocked map
+/// instance
```
Nit: missing a period at the end of the sentence.

illuminati1911 replied:
Fixed.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/example/ios14/ios/RunnerTests/GoogleMapsGroundOverlayControllerTests.m:
```diff
@@ -0,0 +1,155 @@
+}
+
+/// Returns GoogleMapGroundOverlayController object instantiated with bounds and a mocked map
+/// instance
```
Same.

illuminati1911 replied:
Fixed.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/example/ios14/ios/RunnerTests/GoogleMapsGroundOverlayControllerTests.m:
```diff
@@ -0,0 +1,155 @@
+/// instance
+///
+/// @return An object of FLTGoogleMapGroundOverlayController
+- (FGMGroundOverlayController *)groundOverlayControllerWithPositionWithMockedMap {
```
Do these need to be instance methods rather than class methods?

illuminati1911 replied:
Actually no. Changed them to class methods.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/FGMGroundOverlayController.h:
```diff
@@ -0,0 +1,57 @@
+#import <GoogleMaps/GoogleMaps.h>
+
+#import "messages.g.h"
+
```
For new code, the repo preference is to use `NS_ASSUME_NONNULL_*` audited region macros around headers.

illuminati1911 replied:
Added.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/FGMGroundOverlayController.h:
```diff
@@ -0,0 +1,57 @@
+@property(strong, nonatomic) GMSGroundOverlay *_Nonnull groundOverlay;
+
+/// Whether ground overlay is created with bounds or position.
+@property(nonatomic, assign) BOOL isCreatedWithBounds;
```
Properties should not be named "is*"; see https://google.github.io/styleguide/objcguide.html#objective-c-method-names

illuminati1911 replied:
Changed it like this:
```objc
@property(nonatomic, assign, getter=isCreatedWithBounds) BOOL createdWithBounds;
```

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/FGMGroundOverlayController.m:
```diff
@@ -0,0 +1,240 @@
+// found in the LICENSE file.
+
+#import "FGMGroundOverlayController.h"
+#import "FGMUtils.h"
```
Nit: please add a blank line above this; this corresponding header of an implementation file gets its own section.

illuminati1911 replied:
Done.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/FGMGroundOverlayController.m:
```diff
@@ -0,0 +1,240 @@
+
+/// A map from ground overlay id to the controller that manages it.
+@property(strong, nonatomic) NSMutableDictionary<NSString *, FGMGroundOverlayController *>
+    *groundOverlayIdentifierToController;
```
Nit: if you name it `groundOverlayControllerByIdentifier`, then code doing lookups into it will read more like a method, where the last thing names the argument (or in this case, the index).

illuminati1911 replied:
Changed the name.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/FGMGroundOverlayController.m:
```diff
@@ -0,0 +1,240 @@
+}
+
+- (void)didTapGroundOverlayWithIdentifier:(NSString *)identifier {
+  if (!identifier) {
```
Here and in the next method: the identifier params were annotated as non-nullable in the interface, so the implementation code shouldn't be handling nil.

illuminati1911 replied:
Good catch. Removed the null check.

stuartmorgan replied:
It's still here in this method.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/FGMUtils.h:
```diff
@@ -0,0 +1,18 @@
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface FGMUtils : NSObject
```
I strongly discourage "classes" that are just bags of class methods; Objective-C, unlike Java, has a perfectly good way of representing functions that aren't part of a class, which is functions :)

illuminati1911 replied:
My original idea was to use the class for namespacing, but I have now added the functions as global function with FGM prefix.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/FGMUtils.h:
```diff
@@ -0,0 +1,18 @@
+NS_ASSUME_NONNULL_BEGIN
+
+@interface FGMUtils : NSObject
++ (UIImage *)iconFromBitmap:(FGMPlatformBitmap *)platformBitmap
```
These new APIs need declaration comments.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/FGMUtils.h:
Can this be given a more specific name? It's already the case that this plugin has at least one other file of utils (the conversion function file).

illuminati1911 replied:
I renamed it `FGMImageUtils`.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/pigeons/messages.dart:
```diff
@@ -357,6 +357,43 @@ class PlatformCameraTargetBounds {
+  final PlatformBitmap image;
+  final PlatformLatLng? position;
+  final PlatformLatLngBounds? bounds;
+  final PlatformDoublePair? anchor;
```
Isn't this conceptually a point? We already have `PlatformPoint`, so shouldn't need a new `PlatformDoublePair` for this unless there's a conceptual difference in meaning here.

illuminati1911 replied:
Changed this to PlatformPoint.

stuartmorgan reviewed (changes requested):
This looks really good! I did a pass over everything from web, and mostly just have minor comments and questions.

jokerttu commented on packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/ground_overlay.dart:
```diff
@@ -0,0 +1,363 @@
+  /// [MapBitmap.bitmapScaling] must be set to [MapBitmapScaling.none].
+  final MapBitmap image;
+
+  /// Geographical location to which the anchor will be fixed and the [width] of
```
All platforms, Android, iOS and Web differs a bit how GroundOverlays can be intitialized.
Using `bounds` work on each platform, but initializing with position (relative anchor position) is only supported by Android and iOS SDKs. We could also drop position support, and only support bounds if that seems better for unified api, but it would also limit a bit how devs can use the ground overlays when implementing only Android or iOS apps.

The consept of position/anchor, width/height/bounds is a bit hard to explain generally, without having platform specific documentation on platform_interface package.

The `position` of the is geo location of the `anchor` on the map. By default anchor is at position [0.5, 0.5] of the image. So if anchor is not changed and width of the image is given. half of the image is drawn to the west of the anchor position and half if the image to east (on android). If anchor is set to for example [0.0, 0.0] the whole image would be rendered `width`meters amount to the east of the `position`.

Should we add images to the docuementation like this: https://github.com/flutter/flutter/blob/63110570da9fff4a854414fe8910701a82aa3079/packages/flutter/lib/src/painting/box_fit.dart#L31

When initializing ground overlay with position developers need to give: 
-  On Android: `width` of the image in meters on map 
https://developers.google.com/maps/documentation/android-sdk/reference/com/google/android/libraries/maps/model/GroundOverlay
-  On iOS: `zoomLevel` 
https://developers.google.com/maps/documentation/ios-sdk/reference/objc/Classes/GMSGroundOverlay#+groundoverlaywithposition:icon:zoomlevel:

App-facing package has [support table](https://github.com/flutter/packages/pull/8432/files#diff-9bd2e43e18b051bf0a22c1fc3430348ca4b37755cdfd9b4dbb75a98ba156786bR231).

Current implementation also do not support changing the position/bounds/anchor of the groundoverlay after it's created. Not sure is there any use case really when this should be supported. Reason for this is that all platforms allow giving these values properly only on groundoverlay initialization, calculating anchor, position or bounds dynamically from other values could be not accurate. For example anchor can be only given on initialization, if that is changed later, the position would need to be calculated from bounds and updated. This could be implemented in future, if there is really need for it. Current workaround would be creating new groundoverlay instead of moving the existing.

jokerttu replied:
@stuartmorgan updated documentation for the class and its parameters

jokerttu commented on packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/ground_overlay.dart:
```diff
@@ -0,0 +1,363 @@
+  /// the overlay (in meters). The [anchor] is, by default, 50% from the top of
+  /// the image and 50% from the left of the image. This can be changed by
+  /// with the [anchor] parameter.
+  /// You can optionally provide the [height] of the overlay (in meters).
```
Documentation update to clarify how overlays are scaled and placed to the map.

jokerttu commented on packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/ground_overlay.dart:
```diff
@@ -0,0 +1,363 @@
+  final double? height;
+
+  /// Bounds which will contain the image.
+  /// If [bounds] is specified, [position] must be null.
```
Documentation update to clarify how overlays are scaled and placed to the map.

jokerttu commented on packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/ground_overlay.dart:
```diff
@@ -0,0 +1,363 @@
+  /// (0.0, 0.0) means the top left corner of the image. An anchor
+  /// of (1.0, 1.0) means the bottom right corner of the image.
+  /// Defaults to (0.5, 0.5), which is the center of the image.
+  final Offset? anchor;
```
Documentation update to clarify how overlays are scaled and placed to the map.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/example/ios14/ios/RunnerTests/GoogleMapsGroundOverlayControllerTests.m:
```diff
@@ -0,0 +1,156 @@
+
+#import <google_maps_flutter_ios/FGMGroundOverlayController_Test.h>
+
+#import <OCMock/OCMock.h>
```
This should be grouped with the other framework imports. Grouping is by type of thing import, not by `@` vs `#`.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/FGMGroundOverlayController.h:
```diff
@@ -0,0 +1,61 @@
+/// Returns true if a ground overlay with the given identifier exists on the map.
+- (bool)hasGroundOverlaysWithIdentifier:(NSString *)identifier;
+
+/// Returns FGMPlatformGroundOverlay for identifier.
```
"Returns the ground overlay with the given identifier."

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/FGMGroundOverlayController.h:
```diff
@@ -0,0 +1,61 @@
+// found in the LICENSE file.
+
+#import <Flutter/Flutter.h>
+#import <GoogleMaps/GoogleMaps.h>
```
There should be an import of `Foundation` here, as this header uses several Foundation types.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/FGMGroundOverlayController.m:
```diff
@@ -0,0 +1,234 @@
+// found in the LICENSE file.
+
+#import "FGMGroundOverlayController.h"
+
```
This file uses things from UIKit and CoreLocation, so needs to import them.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/FGMGroundOverlayController.m:
```diff
@@ -0,0 +1,234 @@
+                              registrar:(NSObject<FlutterPluginRegistrar> *)registrar
+                            screenScale:(CGFloat)screenScale {
+  [self setConsumeTapEvents:groundOverlay.clickable];
+  [self setVisible:groundOverlay.visible];
```
Based on investigation of similar methods in a recent PR (which hasn't landed), this needs to be last to avoid potential flickering (likely due to Google Maps using its own render thread), so should be at the end of the method, and have a comment saying that nothing should be added after it.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/FGMImageUtils.m:
```diff
@@ -0,0 +1,229 @@
+static UIImage *scaleImage(UIImage *image, double scale);
+static UIImage *scaledImageWithScale(UIImage *image, CGFloat scale);
+static UIImage *scaledImageWithSize(UIImage *image, CGSize size);
+static UIImage *scaledImage(UIImage *image, NSNumber *width, NSNumber *height, CGFloat screenScale);
```
The declaration comments for all of these functions should be here, on the declarations, rather than on the definitions below.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/FGMImageUtils.m:
```diff
@@ -0,0 +1,229 @@
+#import "FGMImageUtils.h"
+#import "Foundation/Foundation.h"
+
+static UIImage *scaleImage(UIImage *image, double scale);
```
This should be `scaled`, not `scale`, for consistency in naming.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/FLTGoogleMapJSONConversions.m:
```diff
@@ -113,6 +113,57 @@ GMSMapViewType FGMGetMapViewTypeForPigeonMapType(FGMPlatformMapType type) {
+                                                    NSNumber *zoomLevel) {
+  /// Dummy image is used as image is required field of FGMPlatformGroundOverlay and converting
+  /// image back to bitmap image is not currently supported.
+  FGMPlatformBitmap *mockImage =
```
This should be called `placeholderImage` rather than `mockImage`, as "mock" generally refers to a specific test construct.

stuartmorgan reviewed (changes requested):
Just some scattered minor issues remaining. Feel free to split out the first sub-PR at this point!

jokerttu commented:
I've addressed the last remaining review comments and split the platform-specific implementations into a separate PR: https://github.com/flutter/packages/pull/8563.
Apologies for the delay!

reidbaker commented on packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/Convert.java:
```diff
@@ -849,6 +850,139 @@ static Tile tileFromPigeon(Messages.PlatformTile tile) {
+   * @return the identifier of the ground overlay.
+   * @throws IllegalArgumentException if required fields are missing or invalid.
+   */
+  static String interpretGroundOverlayOptions(
```
Please add non null or nullable annotations to the objects in this method.

reidbaker commented on packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/Convert.java:
```diff
@@ -849,6 +850,139 @@ static Tile tileFromPigeon(Messages.PlatformTile tile) {
+   *     asset files stored in the application's assets directory.
+   * @param density the density of the display, used to calculate pixel dimensions.
+   * @param wrapper the BitmapDescriptorFactoryWrapper to create BitmapDescriptor.
+   * @return the identifier of the ground overlay.
```
If you know can you add more about the mutability of the ground overlay identifier. For example how long it is a valid id.

reidbaker commented on packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/Convert.java:
```diff
@@ -849,6 +850,139 @@ static Tile tileFromPigeon(Messages.PlatformTile tile) {
+    sink.setClickable(groundOverlay.getClickable());
+    sink.setImage(toBitmapDescriptor(groundOverlay.getImage(), assetManager, density, wrapper));
+    if (groundOverlay.getPosition() != null) {
+      assert groundOverlay.getWidth() != null;
```
Asserts I believe are compiled out so the else condition can still NPE here. Can you convert this assert to a check or handle null width?

reidbaker commented on packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/Convert.java:
```diff
@@ -849,6 +850,139 @@ static Tile tileFromPigeon(Messages.PlatformTile tile) {
+    sink.setImage(toBitmapDescriptor(groundOverlay.getImage(), assetManager, density, wrapper));
+    if (groundOverlay.getPosition() != null) {
+      assert groundOverlay.getWidth() != null;
+      if (groundOverlay.getHeight() != null) {
```
I believe these are the same except the width value. Can you minimize the scope by either making the check cover width only?

reidbaker commented on packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/Convert.java:
```diff
@@ -849,6 +850,139 @@ static Tile tileFromPigeon(Messages.PlatformTile tile) {
+        new Messages.PlatformGroundOverlay.Builder()
+            .setGroundOverlayId(groundOverlayId)
+            .setImage(dummyImage)
+            .setWidth((double) groundOverlay.getWidth())
```
Why use doubles here but floats above for width?

jokerttu replied:
Pigeon uses doubles and google maps uses floats, so conversion between these need to be done while converting values between PlatformGroundOverlay and GoogleMaps SDK GroundOverlay object

reidbaker commented on packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/Convert.java:
```diff
@@ -849,6 +850,139 @@ static Tile tileFromPigeon(Messages.PlatformTile tile) {
+            .setVisible(groundOverlay.isVisible())
+            .setClickable(groundOverlay.isClickable());
+
+    if (isCreatedWithBounds) {
```
It is possible to determine this instead of having it be passed in? For example could a non null getBounds or getPosition indicate that one of those should be used? 

The code above in `interpretGroundOverlayOptions` seem to imply that position is more important that bounds. This code seem to imply that if bounds are used then position should not be.

jokerttu replied:
Google maps object internally have both, bounds and position when the object is created using either of them. When bounds are used, the position representes the achor point in LatLng format for the ground overlay. And if position (with width/height) the bounds are calculated automatically as well.
For integration testing purposes we need track which creation method was used, to be able to check that conversions are handled correctly on each platform.

reidbaker commented on packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/GroundOverlaySink.java:
```diff
@@ -0,0 +1,32 @@
+
+/** Receiver of GroundOverlayOptions configuration. */
+interface GroundOverlaySink {
+  void setTransparency(float transparency);
```
Can you add javadocs to these methods for where someone can learn more about the inputs either here or in groundOverlayController

For example what are the min and max values for transparency. Can zindex be negative? what does setClickable do? etc.

reidbaker commented on packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/GroundOverlaysController.java:
```diff
@@ -0,0 +1,139 @@
+    GroundOverlayController groundOverlayController =
+        groundOverlayIdToController.get(groundOverlayId);
+    if (groundOverlayController != null) {
+      groundOverlayController.remove();
```
Does the order of these removals matter? if so please add a comment explaining the order.

jokerttu replied:
Order does not matter here

reidbaker commented on packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/GroundOverlaysController.java:
```diff
@@ -0,0 +1,139 @@
+    String groundOverlayId = platformGroundOverlay.getGroundOverlayId();
+    GroundOverlayController groundOverlayController =
+        groundOverlayIdToController.get(groundOverlayId);
+    if (groundOverlayController != null) {
```
It is an error to call this with something that cant get a groundOverlayController? If so whos error is it (plugin or host app)?

jokerttu replied:
Api has updateGroundOverlays method takes GroundOverlayUpdates object as parameter:
https://github.com/flutter/packages/blob/ff7724c18a803542fc709f942c3bf1cecf7e4864/packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/ground_overlay_updates.dart
This should be handled by the dart code so that changes only contains items that are available on map.
Not sure would it be good to throw error here if change is triggered for groundoverlay that is not added to the map beforehand. It is indeed misuse of the api, and this is tested in integration tests.

jokerttu commented:
@stuartmorgan This PR now contains only app-facing plugin changes.
Had to update minimum supported SDK versions to Flutter 3.27/Dart 3.6 to match the android package:
https://github.com/flutter/packages/blob/297d5a151fdbabc11a4e0c87d82dca637afa7ecd/packages/google_maps_flutter/google_maps_flutter_android/pubspec.yaml#L7-L9

stuartmorgan reviewed (approved):
LGTM!
